lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,372 @@
+                break;
+            }
+            case TYPE_INT64: {
+                lua_Integer value = luaL_checkinteger(L, -1);
+                bson_array_add_int64(arr, (int64_t)value);
+                break;
+            }
+            case TYPE_STRING: {
+                char *value = strdup(luaL_checkstring(L, -1));
+                bson_array_add_string(arr, value);
+                free(value);
+                break;
+            }
+            case TYPE_DOUBLE: {
+                lua_Number value = luaL_checknumber(L, -1);
+                bson_array_add_double(arr, (double)value);
+                break;
+            }
+            case TYPE_BOOLEAN: {
+                if (!lua_isboolean(L, -1)) {
+                    snprintf(errorMessage, 255, "Expected boolean, but value was of type: %s", 
+                             lua_typename(L, lua_type(L, -1)));
+                    return -1;
+                }
+                int value = lua_toboolean(L, -1);
+
+                bson_array_add_bool(arr, (bson_boolean)value);
+                break;
+            }
+            default: {
+                snprintf(errorMessage, 255, "Error parsing table, unrecognized type: %i", (int)type);
+                return -1;
+            }
+        }
+        lua_pop(L, 2); //Stack: [bson_array]
+    }
+    lua_pop(L, 1); //Stack: []
+    return 0;
+}
+
+static int lbson_object_from_bytes(lua_State *L) {
+    if (!lua_isstring(L, 1)) {
+        return luaL_argerror(L, 1, "Arg 1 must be a Lua String");
+    }
+    size_t len;
+    const char *byte_string = lua_tolstring(L, 1, &len);
+
+    uint8_t *bytes = malloc(len);
+    size_t index = 0;
+    for (index = 0; index < len; index++) {
+        bytes[index] = byte_string[index] & 0xFF;
+    }
+
+    BsonObject obj = bson_object_from_bytes(bytes);
+    free(bytes);
+    
+    char errorMessage[255];
+    int success = bson_object_to_table(L, &obj, errorMessage);
+
+    bson_object_deinitialize(&obj);
+    if (success != 0) {
+        return luaL_argerror(L, 1, errorMessage);
+    }
+    return 1;
+}
+
+static int bson_object_to_table(lua_State *L, BsonObject *obj, char *errorMessage) {
+    MapIterator iter = bson_object_iterator(obj);
+    BsonObjectEntry entry = bson_object_iterator_next(&iter);
+    lua_newtable(L); //Stack: [table]
+    while (entry.element != NULL) {
+        const char *key = entry.key;
+        BsonElement *element = entry.element;
+
+        int success = bson_element_to_table(L, element, errorMessage); //Stack [table, {type, value}]
+        if (success != 0) {
+            return success;
+        }
+
+        lua_setfield(L, -2, key); //Stack [table]
+        entry = bson_object_iterator_next(&iter);
+    }
+    return 0;
+}
+
+static int bson_array_to_table(lua_State *L, BsonArray *arr, char *errorMessage) {
+    lua_newtable(L); //Stack: [table]
+    int index;
+    for (index = 0; index <= arr->count; index++) {
```

Comment:
@jacobkeeler To prevent going out of bound of array in function `bson_array_get()` I propose next change:

`for (index = 0; index <= arr->count; index++) {` 
 -> 
`for (index = 0; index < arr->count; index++)`